controlplane/kubeadm/internal/filters.go

```diff
@@ -29,101 +31,136 @@ import (
 	"sigs.k8s.io/cluster-api/util/collections"
 )
 
-// MatchesMachineSpec returns a filter to find all machines that matches with KCP config and do not require any rollout.
+// matchesMachineSpec returns a filter to find all machines that matches with KCP config and do not require any rollout.
 // Kubernetes version, infrastructure template, and KubeadmConfig field need to be equivalent.
 // Note: We don't need to compare the entire MachineSpec to determine if a Machine needs to be rolled out,
 // because all the fields in the MachineSpec, except for version, the infrastructureRef and bootstrap.ConfigRef, are either:
 // - mutated in-place (ex: NodeDrainTimeout)
 // - are not dictated by KCP (ex: ProviderID)
 // - are not relevant for the rollout decision (ex: failureDomain).
-func MatchesMachineSpec(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
-	return collections.And(
-		collections.MatchesKubernetesVersion(kcp.Spec.Version),
-		MatchesKubeadmBootstrapConfig(machineConfigs, kcp),
-		MatchesTemplateClonedFrom(infraConfigs, kcp),
-	)
+func matchesMachineSpec(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool) {
+	mismatchReasons := []string{}
+
+	if !collections.MatchesKubernetesVersion(kcp.Spec.Version)(machine) {
+		machineVersion := ""
+		if machine != nil && machine.Spec.Version != nil {
+			machineVersion = *machine.Spec.Version
+		}
+		mismatchReasons = append(mismatchReasons, fmt.Sprintf("Machine version %q is not equal to KCP version %q", machineVersion, kcp.Spec.Version))
+	}
+
+	if reason, matches := matchesKubeadmBootstrapConfig(machineConfigs, kcp, machine); !matches {
+		mismatchReasons = append(mismatchReasons, reason)
+	}
+
+	if reason, matches := matchesTemplateClonedFrom(infraConfigs, kcp, machine); !matches {
+		mismatchReasons = append(mismatchReasons, reason)
+	}
+
+	if len(mismatchReasons) > 0 {
+		return fmt.Sprintf("Machine does not match current spec: %s", strings.Join(mismatchReasons, ",")), false
+	}
```

Comment:
What about not nesting messages, it makes them easier to read...
```suggestion
	if len(mismatchReasons) > 0 {
		return strings.Join(mismatchReasons, ","), false
	}
```